File: src/main/java/Utils/Translit.java
package Utils;

import java.util.HashMap;
import java.util.Map;

public class Translit {

    final static int UPPER = 1;

    final static int LOWER = 2;

    final static Map<String, String> map = makeTranslitMap();

    private static Map<String, String> makeTranslitMap() {
        Map<String, String> map = new HashMap<>();
        map.put("a", "а");
        map.put("b", "б");
        map.put("v", "в");
        map.put("g", "г");
        map.put("d", "д");
        map.put("e", "е");
        map.put("yo", "ё");
        map.put("zh", "ж");
        map.put("z", "з");
        map.put("i", "и");
        map.put("j", "й");
        map.put("k", "к");
        map.put("l", "л");
        map.put("m", "м");
        map.put("n", "н");
        map.put("o", "о");
        map.put("p", "п");
        map.put("r", "р");
        map.put("s", "с");
        map.put("t", "т");
        map.put("u", "у");
        map.put("f", "ф");
        map.put("h", "х");
        map.put("ts", "ц");
        map.put("ch", "ч");
        map.put("sh", "ш");
        map.put("`", "ъ");
        map.put("y", "у");
        map.put("'", "ь");
        map.put("yu", "ю");
        map.put("ya", "я");
        map.put("x", "кс");
        map.put("w", "в");
        map.put("q", "к");
        map.put("iy", "ий");
        return map;
    }

    private static int charClass(char c) {
        return Character.isUpperCase(c) ? UPPER : LOWER;
    }

    private static String get(String s) {
        int charClass = charClass(s.charAt(0));
        String result = map.get(s.toLowerCase());
        return result == null ? "" : (charClass == UPPER ? (result.charAt(0) + "").toUpperCase() +
                (result.length() > 1 ? result.substring(1) : "") : result);
    }

    public static String translit(String text) {
        int len = text.length();
        if (len == 0) {
            return text;
        }
        if (len == 1) {
            return get(text);
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < len; ) {
            // get next 2 symbols
            String toTranslate = text.substring(i, i <= len - 2 ? i + 2 : i + 1);
            // trying to translate
            String translated = get(toTranslate);
            // if these 2 symbols are not connected try to translate one by one
            if (translated.isEmpty()) {
                translated = get(toTranslate.charAt(0) + "");
                sb.append(translated.isEmpty() ? toTranslate.charAt(0) : translated);
                i++;
            } else {
                sb.append(translated.isEmpty() ? toTranslate : translated);
                i += 2;
            }
        }
        return sb.toString();
    }
}

File: src/main/kotlin/utils/Translit.kt
package utils

import java.util.HashMap

class Translit {
    final val UPPER:Int=1
    final val LOWER:Int=2
    private val map:Map<String,String>

    constructor(isReverse:Boolean) {
        map = if(isReverse)
            makeReverseTranslitMap()
        else
            makeTranslitMap();
    }

    private fun makeReverseTranslitMap(): Map<String, String> {
        val map:Map<String,String> = mapOf(
                "a" to "а",
                "b" to "б",
                "v" to "в",
                "g" to "г",
                "d" to "д",
                "e" to "е",
                "yo" to "ё",
                "zh" to "ж",
                "z" to "з",
                "i" to "и",
                "j" to "й",
                "k" to "к",
                "l" to "л",
                "m" to "м",
                "n" to "н",
                "o" to "о",
                "p" to "п",
                "r" to "р",
                "s" to "с",
                "t" to "т",
                "u" to "у",
                "f" to "ф",
                "h" to "х",
                "ts" to "ц",
                "ch" to "ч",
                "sh" to "ш",
                "sch" to "щ",
                "\$i" to "ы",
                "`" to "ъ",
                "y" to "у",
                "'" to "ь",
                "yu" to "ю",
                "ya" to "я",
                "x" to "кс",
                "w" to "в",
                "q" to "к"
        )
        return map
    }

    private fun makeTranslitMap():Map<String,String>{
        val map = HashMap<String, String>()
        map["а"] = "a"
        map["б"] = "b"
        map["в"] = "v"
        map["г"] = "g"
        map["д"] = "d"
        map["е"] = "e"
        map["ё"] = "yo"
        map["ж"] = "zh"
        map["з"] = "z"
        map["и"] = "i"
        map["й"] = "j"
        map["л"] = "l"
        map["м"] = "m"
        map["н"] = "n"
        map["о"] = "o"
        map["п"] = "p"
        map["р"] = "r"
        map["с"] = "s"
        map["т"] = "t"
        map["у"] = "u"
        map["ф"] = "f"
        map["х"] = "h"
        map["ц"] = "ts"
        map["ч"] = "ch"
        map["ш"] = "sh"
        map["щ"] = "sch"
        map["ы"] = "\$i"
        map["ъ"] = "`"
        map["ь"] = "'"
        map["ю"] = "yu"
        map["я"] = "ya"
        map["кс"] = "x"
        map["к"] = "q"
        map["ий"] = "iy"
        return map
    }
    private fun charClass(c:Char):Int{
        return if(c.isUpperCase())
            UPPER
        else
            LOWER;
    }

    private fun get(s:String):String{
        val charClass:Int=charClass(s[0])
        val result= map[s.toLowerCase()]
        if(result==null)
            return ""
        else
    }


}